### PR TITLE
layers: Fix object lifetime bugs in DescriptorSetLayout

### PR DESCRIPTION
**Pull request by tobine**
Incomplete copy construction of DescriptorSetLayout and safe_* types
causes embedded pointers (such as pImmutableSamplers) to become invalid
before they should. Fixing this for now by moving DescriptorSetLayout
and its embedded safe_* struct to ptrs within their containers and
explicitly deleting.

### CODE REVIEW

File: layers/descriptor_sets.h [hunk 1]
```diff
@@ -117,7 +117,7 @@ class DescriptorSetLayout {
     unordered_map<uint32_t, uint32_t> binding_to_global_end_index_map_;
     VkDescriptorSetLayoutCreateFlags flags_;
     uint32_t binding_count_; // # of bindings in this layout
-    vector<safe_VkDescriptorSetLayoutBinding> bindings_;
+    vector<safe_VkDescriptorSetLayoutBinding *> bindings_;
```
Comment (lentinem) [inline layers/descriptor_sets.h]: Whats the advantage of using the heap? You could just remove it from the list on deletion which would free the memory.

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -112,7 +112,7 @@ struct layer_data {
     unordered_map<VkCommandPool, CMD_POOL_INFO> commandPoolMap;
     unordered_map<VkDescriptorPool, DESCRIPTOR_POOL_NODE *> descriptorPoolMap;
     unordered_map<VkDescriptorSet, SET_NODE *> setMap;
-    unordered_map<VkDescriptorSetLayout, DescriptorSetLayout> descriptorSetLayoutMap;
+    unordered_map<VkDescriptorSetLayout, DescriptorSetLayout *> descriptorSetLayoutMap;
```
Comment (lentinem) [inline layers/core_validation.cpp]: Same here.

**Comment by karl-lunarg** [comment]
Tobin:  Looks like this fix helped the initial problem, but there seems to be another, later in the CTS:

```
Test case 'dEQP-VK.pipeline.image.view_type.1d.format.r4g4_unorm_pack8.size.1x1'..
Fragment shader compile time = 34.709000 ms
Link time = 0.010000 ms
Vertex shader compile time = 1.200000 ms
Link time = 0.009000 ms
/tmp/hudson4582718127985724795.sh: line 43: 11143 Segmentation fault      (core dumped) ./deqp-vk --deqp-caselist-file=${WORKSPACE}/Vulkan-CTS/external/vulkancts/mustpass/1.0.0/vk-default.txt --deqp-log-images=disable --deqp-log-shader-sources=disable
```

File: layers/descriptor_sets.h (hunk 1 above)
Comment (tobine) [inline layers/descriptor_sets.h]: Without using heap and without complete copy constructor, object immediately goes out of scope when adding to bindings_ vector which causes embedded pImmutableSamplers to be corrupted.

File: layers/core_validation.cpp (hunk 2 above)
Comment (tobine) [inline layers/core_validation.cpp]: same reason.

**Comment by tobine** [comment]
Just force pushed an update that seems to fix CTS bugs as well as pass demos and validation_layer_tests. Please give it a try.

File: layers/descriptor_sets.h (hunk 1 above)
Comment (lentinem) [inline layers/descriptor_sets.h]: I'm not sure I get why. Isn't the idea of safe types to have the ability to have the original objects go out of scope and still store al of the data?

**Comment by karl-lunarg** [comment]
```
Test case 'dEQP-VK.binding_model.shader_access.primary_cmd_buf.uniform_buffer_dynamic.no_access.descriptor_array.offset_view_zero_dynamic_zero'..
Fragment shader compile time = 1.133000 ms
Link time = 0.009000 ms
Vertex shader compile time = 1.216000 ms
Link time = 0.011000 ms
deqp-vk: /var/lib/jenkins/jobs/Vulkan-CTS/workspace/Vulkan-LoaderAndValidationLayers/layers/descriptor_sets.h:191: VkDescriptorType DescriptorSetLayout::GetTypeFromIndex(uint32_t): Assertion `index < bindings_.size()' failed.
/tmp/hudson8715935668804908986.sh: line 43: 23155 Aborted                 (core dumped) ./deqp-vk --deqp-caselist-file=${WORKSPACE}/Vulkan-CTS/external/vulkancts/mustpass/1.0.0/vk-default.txt --deqp-log-images=disable --deqp-log-shader-sources=disable
Build step 'Execute shell' marked build as failure
Finished: FAILURE
```

File: layers/descriptor_sets.h (hunk 1 above)
Comment (tobine) [inline layers/descriptor_sets.h]: Yes, this is just a bug. I'm not generating the copy constructors so need to add that.

**Comment by tobine** [comment]
Added a second commit to this PR that fixes issue with CTS test "dEQP-VK.binding_model.shader_access.primary_cmd_buf.uniform_buffer_dynamic.no_access.descriptor_array.offset_view_zero_dynamic_zero"

**Comment by karl-lunarg** [comment]
CTS passes with this second commit.
Tobin, will you commit to master when able?

**Comment by tobine** [comment]
cherry-picked and pushed to master

**Pull request closed without merging**